Assets/UI/worldinput_CQUI.lua

```diff
@@ -0,0 +1,768 @@
+      cquiHandledKey = true;
+    elseif (unitType == "UNIT_MILITARY_ENGINEER") then
+      CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_FORT"].Hash);
+      cquiHandledKey = true;
+    end
+  end
+
+  if (uiKey == Keys.H and unitType == "UNIT_BUILDER") then
+    CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_CAMP"].Hash);
+    cquiHandledKey = true;
+  end
+
+  if (uiKey == Keys.I and unitType == "UNIT_BUILDER") then
+    CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_FARM"].Hash);
+    cquiHandledKey = true;
+  end
+
+  if (uiKey == Keys.L and unitType == "UNIT_BUILDER") then
+    CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_LUMBER_MILL"].Hash);
+    cquiHandledKey = true;
+  end
+
+  if (uiKey == Keys.N) then
+    cquiHandledKey = true;
+    if (unitType == "UNIT_BUILDER") then
+      CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_MINE"].Hash);
+    else
+      UnitManager.RequestOperation(UI.GetHeadSelectedUnit(), UnitOperationTypes.WMD_STRIKE);
+    end
+  end
+
+  if (uiKey == Keys.O) then
+    if (unitType == "UNIT_BUILDER") then
+      CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_OIL_WELL"].Hash);
+      CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_OFFSHORE_OIL_RIG"].Hash);
+      cquiHandledKey = true;
+    end
+  end
+
+  if (uiKey == Keys.P) then
+    if (CQUI_isShiftDown) then
+      PlaceMapPin();
+      cquiHandledKey = true;
+    else
+      if (unitType == "UNIT_BUILDER") then
+        CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_PASTURE"].Hash);
+        CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_PLANTATION"].Hash);
+        cquiHandledKey = true;
+      end
+    end
+  end
+
+  if (uiKey == Keys.Q) then
+    if (CQUI_isAltDown == true and unitType == "UNIT_BUILDER" ) then
+      CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_QUARRY"].Hash);
+      cquiHandledKey = true;
+    end
+  end
+
+  if (uiKey == Keys.R) then
+    if (unitType == "UNIT_MILITARY_ENGINEER") then
+      -- Build Road/Rail is a UnitOperation
+      CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.UnitOperations["UNITOPERATION_BUILD_ROUTE"].Hash);
+      cquiHandledKey = true;
+    elseif (formationClass == "FORMATION_CLASS_AIR") then
+      UnitManager.RequestOperation(UI.GetHeadSelectedUnit(), UnitOperationTypes.REBASE);
+      cquiHandledKey = true;
+    end
+  end
+
+  if (uiKey == Keys.S and CQUI_isAltDown == true) then
+    if (formationClass == "FORMATION_CLASS_AIR") then
+      UnitManager.RequestCommand(UI.GetHeadSelectedUnit(), UnitOperationTypes.AIR_ATTACK);
+      cquiHandledKey = true;
+    end
```

Comment:
Can't seem to get either of these air unit commands to work. Did it work for you? Also theoretically the S key one should be `RequestOperation` function right? I tried it with that and it also doesn't work.

Reply:
I was looking through the unmodded files for anything related to air unit rebasing. Found these functions `OnMouseRebaseEnd`, `AirUnitReBase` and `OnInterfaceModeChange_ReBase` in the WorldInput.lua file. Also the `RequestOperation` call looks like this
`UnitManager.RequestOperation( pSelectedUnit, UnitOperationTypes.REBASE, tParameters);`.

The stuff related to air attack also seems to be a multi-stage, multi-function process. `UnitAirAttack`, `OnInterfaceModeChange_Air_Attack`.

The WMD stuff also looks like it is multi-stage process, but I haven't tried it yet. Did you try the hotkey? Based on what I'm seeing in the base game code, it shouldn't work since it also requires a third parameter which we don't provide.

Right now I'm thinking it might not be a simple single command that we can issue, if that is the case maybe we should just remove it from this pull and work on it separately.

Reply:
I opened up Issue #40 to track this.  I'll add your note there, as well.  Commented out the code in the KeyUp handler in the meantime, with a note.

At this point I want to get the fix for the Soothsayer done and think it's an okay tradeoff to lose these hotkeys that were not likely working as intended anyway